File: geneo/lib/identify.cpp
/*------------------------------------------------------------------------------
MIT License

Copyright (c) 2024 Gilles-Philippe Morin

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
------------------------------------------------------------------------------*/

#include "../include/identify.hpp"

// Returns the founder IDs of a pedigree.
std::vector<int> get_founder_ids(Pedigree<> &pedigree) {
    std::vector<int> founder_ids;
    for (auto const& [id, individual] : pedigree.individuals) {
        if (!individual->father && !individual->mother) {
            founder_ids.push_back(id);
        }
    }
    // Sort the founders by ID
    sort(founder_ids.begin(), founder_ids.end());
    return founder_ids;
}

// Returns the half founder IDs of a pedigree.
std::vector<int> get_half_founder_ids(Pedigree<> &pedigree) {
    std::vector<int> half_founder_ids;
    for (auto const& [id, individual] : pedigree.individuals) {
        if ((!individual->father && individual->mother) ||
            (individual->father && !individual->mother)) {
            half_founder_ids.push_back(id);
        }
    }
    // Sort the half founders by ID
    sort(half_founder_ids.begin(), half_founder_ids.end());
    return half_founder_ids;
}

// Returns the proband IDs of a pedigree.
std::vector<int> get_proband_ids(Pedigree<> &pedigree) {
    std::vector<int> proband_ids;
    for (auto const& [id, individual] : pedigree.individuals) {
        if (!individual->children.size()) {
            proband_ids.push_back(id);
        }
    }
    // Sort the probands by ID
    sort(proband_ids.begin(), proband_ids.end());
    return proband_ids;
}

// Returns the IDs of the fathers of a vector of individuals.
std::vector<int> get_father_ids(
    Pedigree<> &pedigree, std::vector<int> ids
) {
    std::vector<int> father_ids;
    std::set<int> set;
    for (const int id : ids) {
        Individual<> *individual = pedigree.individuals.at(id);
        if (individual->father) {
            set.insert(individual->father->id);
        }
    }
    for (const int id : set) {
        father_ids.push_back(id);
    }
    // Sort the fathers by ID
    sort(father_ids.begin(), father_ids.end());
    return father_ids;
}

// Returns the IDs of the mothers of a vector of individuals.
std::vector<int> get_mother_ids(
    Pedigree<> &pedigree, std::vector<int> ids
) {
    std::vector<int> mother_ids;
    std::set<int> set;
    for (const int id : ids) {
        Individual<> *individual = pedigree.individuals.at(id);
        if (individual->mother) {
            set.insert(individual->mother->id);
        }
    }
    for (const int id : set) {
        mother_ids.push_back(id);
    }
    // Sort the mothers by ID
    sort(mother_ids.begin(), mother_ids.end());
    return mother_ids;
}

// Returns the IDs of the siblings of a vector of individuals.
std::vector<int> get_sibling_ids(Pedigree<> &pedigree,
    std::vector<int> ids, bool include_half_siblings) {
    std::vector<int> sibling_ids;
    std::set<int> set1, set2, result;
    for (const int id : ids) {
        set2.insert(id);
        std::set<int> fathers_children, mothers_children;
        Individual<> *individual = pedigree.individuals.at(id);
        if (individual->father) {
            for (Individual<> *child : individual->father->children) {
                fathers_children.insert(child->id);
            }
        }
        if (individual->mother) {
            for (Individual<> *child : individual->mother->children) {
                mothers_children.insert(child->id);
            }
        }
        if (include_half_siblings) {
            set_union(
                fathers_children.begin(), fathers_children.end(),
                mothers_children.begin(), mothers_children.end(),
                std::inserter(set1, set1.begin())
            );
        } else {
            set_intersection(
                fathers_children.begin(), fathers_children.end(),
                mothers_children.begin(), mothers_children.end(),
                std::inserter(set2, set2.begin())
            );
        }
    }
    set_difference(
        set1.begin(), set1.end(),
        set2.begin(), set2.end(),
        std::inserter(result, result.begin())
    );
    for (const int id : result) {
        sibling_ids.push_back(id);
    }
    // Sort the siblings by ID
    sort(sibling_ids.begin(), sibling_ids.end());
    return sibling_ids;
}

// Returns the IDs of the children of a vector of individuals.
std::vector<int> get_children_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> children_ids;
    std::unordered_set<int> set;
    for (int id : ids) {
        Individual<> *individual = pedigree.individuals.at(id);
        for (Individual<> *child : individual->children) {
            set.insert(child->id);
        }
    }
    for (int id : set) {
        children_ids.push_back(id);
    }
    // Sort the children by ID
    sort(children_ids.begin(), children_ids.end());
    return children_ids;
}

// Returns the IDs of the ancestors of a vector of individuals.
std::vector<int> get_ancestor_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> ancestor_ids;
    std::unordered_set<int> set;
    std::deque<int> queue;
    for (const int id : ids) {
        queue.push_back(id);
    }
    while (!queue.empty()) {
        const int id = queue.front();
        queue.pop_front();
        Individual<> *individual = pedigree.individuals.at(id);
        if (individual->father) {
            set.insert(individual->father->id);
            queue.push_back(individual->father->id);
        }
        if (individual->mother) {
            set.insert(individual->mother->id);
            queue.push_back(individual->mother->id);
        }
    }
    for (const int id : set) {
        ancestor_ids.push_back(id);
    }
    // Sort the ancestors by ID
    sort(ancestor_ids.begin(), ancestor_ids.end());
    return ancestor_ids;
}

// Returns the IDs of all occurrences of ancestors of a vector of individuals.
std::vector<int> get_all_ancestor_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> ancestor_ids;
    std::deque<int> queue;
    for (const int id : ids) {
        queue.push_back(id);
    }
    while (!queue.empty()) {
        const int id = queue.front();
        queue.pop_front();
        Individual<> *individual = pedigree.individuals.at(id);
        if (individual->father) {
            ancestor_ids.push_back(individual->father->id);
            queue.push_back(individual->father->id);
        }
        if (individual->mother) {
            ancestor_ids.push_back(individual->mother->id);
            queue.push_back(individual->mother->id);
        }
    }
    // Sort the ancestors by ID
    sort(ancestor_ids.begin(), ancestor_ids.end());
    return ancestor_ids;
}

// Returns the IDs of the descendants of a vector of individuals.
std::vector<int> get_descendant_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> descendant_ids;
    std::unordered_set<int> set;
    std::deque<int> queue;
    for (const int id : ids) {
        queue.push_back(id);
    }
    while (!queue.empty()) {
        const int id = queue.front();
        queue.pop_front();
        Individual<> *individual = pedigree.individuals.at(id);
        for (Individual<> *child : individual->children) {
            set.insert(child->id);
            queue.push_back(child->id);
        }
    }
    for (const int id : set) {
        descendant_ids.push_back(id);
    }
    // Sort the descendants by ID
    sort(descendant_ids.begin(), descendant_ids.end());
    return descendant_ids;
}

// Returns the IDs of all occurrences of descendants of a vector of individuals.
std::vector<int> get_all_descendant_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> descendant_ids;
    std::deque<int> queue;
    for (const int id : ids) {
        queue.push_back(id);
    }
    while (!queue.empty()) {
        const int id = queue.front();
        queue.pop_front();
        Individual<> *individual = pedigree.individuals.at(id);
        for (Individual<> *child : individual->children) {
            descendant_ids.push_back(child->id);
            queue.push_back(child->id);
        }
    }
    // Sort the descendants by ID
    sort(descendant_ids.begin(), descendant_ids.end());
    return descendant_ids;
}

// Returns the IDs of the common ancestors of a vector of individuals.
std::vector<int> get_common_ancestor_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> common_ancestor_ids = get_ancestor_ids(pedigree, ids);
    std::set<int> set1, set2;
    std::vector<int> intersection_result;
    for (int id : common_ancestor_ids) {
        set1.insert(id);
    }
    common_ancestor_ids.clear();
    for (int id : ids) {
        std::vector<int> ancestor_ids = get_ancestor_ids(
            pedigree, std::vector<int>(1, id)
        );
        for (int ancestor_id : ancestor_ids) {
            set2.insert(ancestor_id);
        }
        set_intersection(
            set1.begin(), set1.end(),
            set2.begin(), set2.end(),
            std::back_inserter(intersection_result)
        );
        set1.clear();
        for (const int id : intersection_result) {
            set1.insert(id);
        }
        set2.clear();
        intersection_result.clear();
    }
    for (int id : set1) {
        common_ancestor_ids.push_back(id);
    }
    // Sort the common ancestors by ID
    sort(common_ancestor_ids.begin(), common_ancestor_ids.end());
    return common_ancestor_ids;
}

// Returns the IDs of the common founders of a vector of individuals.
std::vector<int> get_common_founder_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> common_ancestor_ids =
        get_common_ancestor_ids(pedigree, ids);
    std::set<int> set1(common_ancestor_ids.begin(), common_ancestor_ids.end());
    std::vector<int> founder_ids = get_founder_ids(pedigree);
    std::set<int> set2(founder_ids.begin(), founder_ids.end());
    std::set<int> intersection_result;
    set_intersection(
        set2.begin(), set2.end(),
        set1.begin(), set1.end(),
        std::inserter(intersection_result, intersection_result.begin())
    );
    std::vector<int> common_founder_ids(
        intersection_result.begin(),
        intersection_result.end()
    );
    // Sort the common founders by ID
    sort(common_founder_ids.begin(), common_founder_ids.end());
    return common_founder_ids;
}

// Returns the IDs of the most recent common ancestors of a vector of individuals.
std::vector<int> get_mrca_ids(Pedigree<> &pedigree,
    std::vector<int> ids) {
    std::vector<int> most_recent_common_ancestor_ids;
    std::vector<int> common_ancestor_ids =
        get_common_ancestor_ids(pedigree, ids);
    std::vector<int> ancestors_of_common_ancestors =
        get_ancestor_ids(pedigree, common_ancestor_ids);
    std::set<int> set1(common_ancestor_ids.begin(), common_ancestor_ids.end());
    std::set<int> set2(
        ancestors_of_common_ancestors.begin(),
        ancestors_of_common_ancestors.end()
    );
    set_difference(
        set1.begin(), set1.end(),
        set2.begin(), set2.end(),
        std::back_inserter(most_recent_common_ancestor_ids)
    );
    // Sort the most recent common ancestors by ID
    sort(
        most_recent_common_ancestor_ids.begin(), 
        most_recent_common_ancestor_ids.end()
    );
    return most_recent_common_ancestor_ids;
}

File: geneo/identify.py
import numpy as np
import pandas as pd
import cgeneo

def pro(gen):
    """Get proband IDs (individuals without children in the genealogy)
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        
    Returns:
        list: Proband IDs sorted in ascending order
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaJi
        >>> ped = gen.genealogy(geneaJi)
        >>> pro = gen.pro(pedigree)
        >>> print(pro)
        [1, 2, 29]
        
    See Also:
        gen.founder: Get individuals without parents
        gen.children: Find children of specified individuals
    """
    proband_ids = cgeneo.get_proband_ids(gen)
    return proband_ids

def founder(gen):
    """Get founder IDs (individuals without known parents)
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
    
    Returns:
        list: Founder IDs sorted in ascending order
        
    Notes:
        Founders are defined as individuals with both parents = 0
        Use gen.half_founder() for individuals with one unknown parent
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(genea140)
        >>> founders = gen.founder(pedigree)
        >>> print(f"Founders count: {len(founders)}")
        Founders count: 7399
    """
    founder_ids = cgeneo.get_founder_ids(gen)
    return founder_ids

def half_founder(gen):
    """Get half-founder IDs (individuals with one unknown parent)
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        
    Returns:
        list: Half-founder IDs sorted in ascending order
        
    Notes:
        Half-founders have either father=0 or mother=0, but not both
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaJi
        >>> ped = gen.genealogy(geneaJi)
        >>> hf = gen.half_founder(pedigree)
        >>> print(hf)
        [9, 11]
    """
    half_founder_ids = cgeneo.get_half_founder_ids(gen)
    return half_founder_ids

def parent(gen, individuals, **kwargs):
    """Get parental IDs for specified individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        output (str): Parent type to return:
            'FaMo' - Both parents (default)
            'Fa' - Fathers only
            'Mo' - Mothers only
            
    Returns:
        dict: Dictionary with 'Fathers' and/or 'Mothers' keys containing
              lists of parental IDs (0 indicates unknown parent)
              
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaJi
        >>> ped = gen.genealogy(geneaJi)
        >>> pro = gen.pro(ped)
        >>> parents = gen.parent(ped, pro)
        >>> print(parents['Fathers'])
        [4, 28]
        
    See Also:
        gen.children: Get inverse relationship
        gen.sibship: Find siblings through shared parents
    """
    output = kwargs.get('output', 'FaMo')
    parents = dict()
    if output == 'FaMo':
        parents['Fathers'] = cgeneo.get_father_ids(gen, individuals)
        parents['Mothers'] = cgeneo.get_mother_ids(gen, individuals)
    if output == 'Fa':
        parents['Fathers'] = cgeneo.get_father_ids(gen, individuals)
    if output == 'Mo':
        parents['Mothers'] = cgeneo.get_mother_ids(gen, individuals)
    return parents

def sibship(gen, individuals, **kwargs):
    """Get sibling IDs for specified individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        halfSibling (bool): Include half-siblings if True (default)
        
    Returns:
        dict: Dictionary mapping each individual to their siblings' IDs
        
    Notes:
        Full siblings share both parents
        Half-siblings share one parent
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(genea140)
        >>> siblings = gen.sibship(ped, [113470])
        >>> print(siblings)
        [10033]
    """
    halfSibling = kwargs.get('halfSibling', True)
    sibling_ids = cgeneo.get_sibling_ids(gen, individuals, halfSibling)
    return sibling_ids

def children(gen, individuals):
    """Get children IDs for specified individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        
    Returns:
        list: Children's IDs sorted in alphabetical order
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(genea140)
        >>> children = gen.children(ped, [10086])
        >>> print(children)
        [33724]
        
    See Also:
        gen.parent: Inverse relationship lookup
    """
    children_ids = cgeneo.get_children_ids(gen, individuals)
    return children_ids

def ancestor(gen, individuals, **kwargs):
    """Get ancestor IDs for specified individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        type (str): Ancestor retrieval mode:
            'UNIQUE' - One occurrence per ancestor (default)
            'TOTAL' - All occurrences of ancestors
            
    Returns:
        list: Ancestor IDs
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(genea140)
        >>> pro = gen.pro(ped)
        >>> all_ancestors = gen.ancestor(ped, pro, type='TOTAL')
        >>> print(f"Ancestors count: {len(all_ancestors)}")
        Ancestors count: 575418
    """
    type = kwargs.get('type', 'UNIQUE')
    if type == 'UNIQUE':
        ancestor_ids = cgeneo.get_ancestor_ids(gen, individuals)
    elif type == 'TOTAL':
        ancestor_ids = cgeneo.get_all_ancestor_ids(gen, individuals)
    return ancestor_ids

def descendant(gen, individuals, **kwargs):
    """Get descendant IDs for specified individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        type (str): Descendant retrieval mode:
            'UNIQUE' - One occurrence per descendant (default)
            'TOTAL' - All occurrences of descendants
            
    Returns:
        list: Descendant IDs
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaa140
        >>> ped = gen.genealogy(genea140)
        >>> founders = gen.founder(ped)
        >>> all_descendants = gen.descendant(ped, founders, type='TOTAL')
        >>> print(f"Descendants count: {len(all_descendants)}")
        Descendants count: 2123427
    """
    type = kwargs.get('type', 'UNIQUE')
    if type == 'UNIQUE':
        descendant_ids = cgeneo.get_descendant_ids(gen, individuals)
    elif type == 'TOTAL':
        descendant_ids = cgeneo.get_all_descendant_ids(gen, individuals)
    return descendant_ids

def commonAncestor(gen, individuals):
    """Find common ancestors shared by multiple individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        
    Returns:
        list: Shared ancestor IDs sorted in ascending order
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(genea140)
        >>> common_anc = gen.commonAncestor(ped, [113470, 10033])
        >>> print(f"Common ancestors count: {len(common_anc)}")
        Common ancestors count: 38
    """
    common_ancestor_ids = cgeneo.get_common_ancestor_ids(gen, individuals)
    return common_ancestor_ids

def findFounders(gen, individuals):
    """Find common founders shared by multiple individuals
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        
    Returns:
        list: Shared founder IDs sorted in ascending order
        
    Examples:
        >>> import geneo as gen
        >>> from geneo import genea140
        >>> ped = gen.genealogy(geneaJi)
        >>> shared_founders = gen.findFounders(ped, [1, 29])
        >>> print(len(shared_founders))
        5
    """
    common_founder_ids = cgeneo.get_common_founder_ids(gen, individuals)
    return common_founder_ids

def findMRCA(gen, individuals):
    """Identify Most Recent Common Ancestors (MRCAs)
    
    Args:
        gen (cgeneo.Pedigree): Initialized genealogy object
        individuals (list): Target individual IDs
        
    Returns:
        pd.DataFrame: Meioses matrix with:
            - Rows: Target individuals
            - Columns: MRCA IDs
            - Values: Number of meioses (generational steps)
            
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaJi
        >>> ped = gen.genealogy(geneaJi)
        >>> mrca_matrix = gen.findMRCA(ped, [1, 29])
        >>> print(mrca_matrix)
            14  20
        1    4   4
        29   3   3
    """
    mrca_ids = cgeneo.get_mrca_ids(gen, individuals)
    cmatrix = cgeneo.get_mrca_meioses(gen, individuals, mrca_ids)
    meioses_matrix = pd.DataFrame(
        cmatrix, index=individuals, columns=mrca_ids, copy = False)
    return meioses_matrix

def find_Min_Distance_MRCA(genMatrix, **kwargs):
    """Calculate minimum genetic distances through MRCAs
    
    Args:
        genMatrix (pd.DataFrame): Meioses matrix from findMRCA
        individuals (list): Subset of individuals to analyze (default: all)
        ancestors (list): Subset of ancestors to consider (default: all)
        
    Returns:
        pd.DataFrame: Minimum distances with columns:
            - founder: MRCA ID
            - proband1: First individual ID
            - proband2: Second individual ID  
            - distance: Total meioses count
            
    Examples:
        >>> import geneo as gen
        >>> from geneo import geneaJi
        >>> ped = gen.genealogy(geneaJi)
        >>> mrca_matrix = gen.findMRCA(ped, [1, 29])
        >>> min_dist = gen.find_Min_Distance_MRCA(mrca_matrix)
        >>> print(min_dist)
           founder  proband1  proband2  distance
        0       14         1        29         4
        1       20         1        29         4
    """
    individuals = kwargs.get('individuals', None)
    if individuals is None:
        individuals = genMatrix.index
    ancestors = kwargs.get('ancestors', None)
    if ancestors is None:
        ancestors = genMatrix.columns
        
    meioses_matrix = genMatrix.loc[individuals, ancestors]
    founder_vector = []
    proband1_vector = []
    proband2_vector = []
    distances_vector = []
    
    for ancestor in ancestors:
        min_distance = np.inf
        min_pairs = []
        for i, proband1 in enumerate(individuals):
            for proband2 in individuals[i+1:]:
                distance = meioses_matrix.at[proband1, ancestor] 
                + meioses_matrix.at[proband2, ancestor]
                if distance < min_distance:
                    min_distance = distance
                    min_pairs = [(proband1, proband2)]
                elif distance == min_distance:
                    min_pairs.append((proband1, proband2))
        for proband1, proband2 in min_pairs:
            founder_vector.append(ancestor)
            proband1_vector.append(proband1)
            proband2_vector.append(proband2)
            distances_vector.append(min_distance)
            
    df = pd.DataFrame({
        'founder': founder_vector,
        'proband1': proband1_vector,
        'proband2': proband2_vector,
        'distance': distances_vector
    })
    return df